## For Part 1:

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
!pip install lxml

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


In [3]:
listt = pd.read_html("https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M")

In [4]:
df = listt[0]
df.head()

Postal Code           Borough               Neighborhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

Let us only process the cells that have an assigned borough.

In [5]:
df = df[df['Borough'] != 'Not assigned']
df.reset_index(drop=True, inplace=True)
df.head()

Postal Code           Borough                                 Neighborhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

Let us check whether there is any 'Not assigned' neighborhood or not

In [6]:
df1 = df[df['Neighborhood'] == 'Not assigned']
df1.head()

Empty DataFrame
Columns: [Postal Code, Borough, Neighborhood]
Index: []

So it seems there is not any 'Not assigned' neighborhood

In [7]:
df.shape

(103, 3)

There are 103 rows and 3 columns in my dataframe df.

## Part 2:

In [8]:
df_lat_long = pd.read_csv('http://cocl.us/Geospatial_data')
df_lat_long.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [12]:
df_new = pd.merge(df, df_lat_long, on='Postal Code')
df_new.head()

Postal Code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

In [13]:
df_new.shape

(103, 5)

In [14]:
print('The dataframe has {} boroughs and at least {} neighborhoods.'.format(
        len(df_new['Borough'].unique()),
        len(df_new['Neighborhood'].unique())
    )
)

The dataframe has 10 boroughs and at least 99 neighborhoods.


## Part 3:

Let us use geopy library to get the latitude and longitude values of Toronto.

In [15]:
address = 'Toronto, CA'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


In [16]:
# create map of New York using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_new['Latitude'], df_new['Longitude'], df_new['Borough'], df_new['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

However, for illustration purposes, let's simplify the above map and segment and cluster only the neighborhoods in Downtown Toronto. So let's slice the original dataframe and create a new dataframe of the Downtown Toronto data.

In [17]:
down_data = df_new[df_new['Borough'] == 'Downtown Toronto'].reset_index(drop=True)
down_data.head()

Postal Code           Borough                                 Neighborhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M5E  Downtown Toronto                                  Berczy Park   

    Latitude  Longitude  
0  43.654260 -79.360636  
1  43.662301 -79.389494  
2  43.657162 -79.378937  
3  43.651494 -79.375418  
4  43.644771 -79.373306

Let's get the geographical coordinates of Downtown Toronto.

In [18]:
address = 'Downtown Toronto, CA'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Downtown Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Downtown Toronto are 43.6563221, -79.3809161.


As we did with all of Toronto, let's visualize  Downtown Toronto the neighborhoods in it.

In [19]:
# create map of Manhattan using latitude and longitude values
map_downtown_to = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(down_data['Latitude'], down_data['Longitude'], down_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_downtown_to)  
    
map_downtown_to

In [20]:
CLIENT_ID = 'WKBRV1KESRPMV0IDVVG243W4TSD32542XD3SJY4N5YOYS430' # your Foursquare ID
CLIENT_SECRET = 'WEWY2SFEBSFDFO3COMU5V4EBDQAF1YCUMPQGR1SISCUQA4IS' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version


 Let's explore the first neighborhood in our dataframe.

Get the neighborhood's name.

In [21]:
down_data.loc[0, 'Neighborhood']

'Regent Park, Harbourfront'

In [22]:
neighborhood_latitude = down_data.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = down_data.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = down_data.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Regent Park, Harbourfront are 43.6542599, -79.3606359.


Now, let's get the top 100 venues that are in Regent Park, Harbourfront within a radius of 500 meters.

First, let's create the GET request URL. Name your URL **url**.

In [23]:
# type your answer here
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius

url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url


'https://api.foursquare.com/v2/venues/explore?&client_id=WKBRV1KESRPMV0IDVVG243W4TSD32542XD3SJY4N5YOYS430&client_secret=WEWY2SFEBSFDFO3COMU5V4EBDQAF1YCUMPQGR1SISCUQA4IS&v=20180605&ll=43.6542599,-79.3606359&radius=500&limit=100'

In [24]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5eeb4ddbedbcad001bfa9dcf'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Corktown',
  'headerFullLocation': 'Corktown, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 44,
  'suggestedBounds': {'ne': {'lat': 43.6587599045, 'lng': -79.3544279001486},
   'sw': {'lat': 43.6497598955, 'lng': -79.36684389985142}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '54ea41ad498e9a11e9e13308',
       'name': 'Roselle Desserts',
       'location': {'address': '362 King St E',
        'crossStreet': 'Trinity St',
        'lat': 43.653446723052674,
        'lng': -79.3620167174383,
        'labeledLatLngs': [{'label': 'display',
 

From the Foursquare lab in the previous module, we know that all the information is in the items key. Before we proceed, let's borrow the get_category_type function from the Foursquare lab.

In [25]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Now we are ready to clean the json and structure it into a pandas dataframe.

In [26]:
venues = results['response']['groups'][0]['items']
venues  

[{'reasons': {'count': 0,
   'items': [{'summary': 'This spot is popular',
     'type': 'general',
     'reasonName': 'globalInteractionReason'}]},
  'venue': {'id': '54ea41ad498e9a11e9e13308',
   'name': 'Roselle Desserts',
   'location': {'address': '362 King St E',
    'crossStreet': 'Trinity St',
    'lat': 43.653446723052674,
    'lng': -79.3620167174383,
    'labeledLatLngs': [{'label': 'display',
      'lat': 43.653446723052674,
      'lng': -79.3620167174383}],
    'distance': 143,
    'postalCode': 'M5A 1K9',
    'cc': 'CA',
    'city': 'Toronto',
    'state': 'ON',
    'country': 'Canada',
    'formattedAddress': ['362 King St E (Trinity St)',
     'Toronto ON M5A 1K9',
     'Canada']},
   'categories': [{'id': '4bf58dd8d48988d16a941735',
     'name': 'Bakery',
     'pluralName': 'Bakeries',
     'shortName': 'Bakery',
     'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/bakery_',
      'suffix': '.png'},
     'primary': True}],
   'photos': {'count': 0, 'grou

In [27]:
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/ipykernel_launcher.py:1: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  """Entry point for launching an IPython kernel.


name           categories        lat        lng
0        Roselle Desserts               Bakery  43.653447 -79.362017
1           Tandem Coffee          Coffee Shop  43.653559 -79.361809
2  Cooper Koo Family YMCA  Distribution Center  43.653249 -79.358008
3     Body Blitz Spa East                  Spa  43.654735 -79.359874
4         Corktown Common                 Park  43.655618 -79.356211

In [28]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

44 venues were returned by Foursquare.


### Explore Neighborhoods in Downtown Toronto

#### Let's create a function to repeat the same process to all the neighborhoods in Downtown Toronto

In [29]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Now write the code to run the above function on each neighborhood and create a new dataframe called Down_to_venues.

In [32]:

down_to_venues = getNearbyVenues(names=down_data['Neighborhood'],
                                   latitudes=down_data['Latitude'],
                                   longitudes=down_data['Longitude']
                                  )

Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Harbourfront East, Union Station, Toronto Islands
Toronto Dominion Centre, Design Exchange
Commerce Court, Victoria Hotel
University of Toronto, Harbord
Kensington Market, Chinatown, Grange Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Rosedale
Stn A PO Boxes
St. James Town, Cabbagetown
First Canadian Place, Underground city
Church and Wellesley


#### Let's check the size of the resulting dataframe

In [33]:
print(down_to_venues.shape)
down_to_venues.head()

(1227, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront               43.65426              -79.360636   
1  Regent Park, Harbourfront               43.65426              -79.360636   
2  Regent Park, Harbourfront               43.65426              -79.360636   
3  Regent Park, Harbourfront               43.65426              -79.360636   
4  Regent Park, Harbourfront               43.65426              -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2  Cooper Koo Family YMCA       43.653249       -79.358008   
3     Body Blitz Spa East       43.654735       -79.359874   
4         Corktown Common       43.655618       -79.356211   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2  Distribution Center  
3                  Spa  
4                 Park

Let's check how many venues were returned for each neighborhood

In [34]:
down_to_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            58   
CN Tower, King and Spadina, Railway Lands, Harb...                     16   
Central Bay Street                                                     67   
Christie                                                               17   
Church and Wellesley                                                   80   
Commerce Court, Victoria Hotel                                        100   
First Canadian Place, Underground city                                100   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
Kensington Market, Chinatown, Grange Park                              61   
Queen's Park, Ontario Provincial Government                            32   
Regent Park, Harbourfront                                              44   
Richmond, Adelaide, King                                               93   
Rosedale                                                                4   
St. James Town                                                         78   
St. James Town, Cabbagetown                                            45   
Stn A PO Boxes                                                         97   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         35   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             58   
CN Tower, King and Spadina, Railway Lands, Harb...                      16   
Central Bay Street                                                      67   
Christie                                                                17   
Church and Wellesley                                                    80   
Commerce Court, Victoria Hotel                                         100   
First Canadian Place, Underground city                                 100   
Garden District, Ryerson                                               100   
Harbourfront East, Union Station, Toronto Islands                      100   
Kensington Market, Chinatown, Grange Park                               61   
Queen's Park, Ontario Provincial Government                             32   
Regent Park, Harbourfront                                               44   
Richmond, Adelaide, King                                                93   
Rosedale                                                                 4   
St. James Town                                                          78   
St. James Town, Cabbagetown                                             45   
Stn A PO Boxes                                                          97   
Toronto Dominion Centre, Design Exchange                               100   
University of Toronto, Harbord                                          35   

                                                    Venue  Venue Latitude  \
Neighborhood                                                                
Berczy Park                                            58              58   
CN Tower, King and Spadina, Railway Lands, Harb...     16              16   
Central Bay Street                                     67              67   
Christie                                               17              17   
Church and Wellesley                                   80              80   
Commerce Court, Victoria Hotel                        100             100   
First Canadian Place, Underground city                100             100   
Garden District, Ryerson                              100             100   
Harbourfront East, Union 

#### Let's find out how many unique categories can be curated from all the returned venues

In [35]:
print('There are {} uniques categories.'.format(len(down_to_venues['Venue Category'].unique())))

There are 214 uniques categories.


##  Analyze Each Neighborhood

In [36]:
# one hot encoding
down_to_onehot = pd.get_dummies(down_to_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
down_to_onehot['Neighborhood'] = down_to_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [down_to_onehot.columns[-1]] + list(down_to_onehot.columns[:-1])
down_to_onehot = down_to_onehot[fixed_columns]

down_to_onehot.head()

Yoga Studio  Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0            0                  0        0                   0             0   
1            0                  0        0                   0             0   
2            0                  0        0                   0             0   
3            0                  0        0                   0             0   
4            0                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0             0         0            0           0                    0   
1             0         0            0           0                    0   
2             0         0            0           0                    0   
3             0         0            0           0                    0   
4             0         0            0           0                    0   

   Asian Restaurant  Athletics & Sports  BBQ Joint  Baby Store  Bagel Shop  \
0                 0                   0          0           0           0   
1                 0                   0          0           0           0   
2                 0                   0          0           0           0   
3                 0                   0          0           0           0   
4                 0                   0          0           0           0   

   Bakery  Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  \
0       1     0    0                 0                   0      0   
1       0     0    0                 0                   0      0   
2       0     0    0                 0                   0      0   
3       0     0    0                 0                   0      0   
4       0     0    0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bistro  \
0                0         0           0                   0       0   
1                0         0           0                   0       0   
2                0         0           0                   0       0   
3                0         0           0                   0       0   
4                0         0           0                   0       0   

   Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  \
0              0          0         0                     0               0   
1              0          0         0                     0               0   
2              0          0         0                     0               0   
3              0          0         0                     0               0   
4              0          0         0                     0               0   

   Brewery  Bubble Tea Shop  Building  Burger Joint  Burrito Place  Butcher  \
0        0                0         0             0              0        0   
1        0                0         0             0              0        0   
2        0                0         0             0              0        0   
3        0                0         0             0              0        0   
4        0                0         0             0              0        0   

   Cable Car  Café  Camera Store  Candy Store  Caribbean Restaurant  \
0          0     0             0            0                     0   
1          0     0             0            0                     0   
2          0     0             0            0                     0   
3          0     0             0            0            

In [37]:
down_to_onehot.shape

(1227, 214)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [38]:
down_to_grouped = down_to_onehot.groupby('Neighborhood').mean().reset_index()
down_to_grouped.head(3)

Neighborhood  Yoga Studio  \
0                                        Berczy Park     0.000000   
1  CN Tower, King and Spadina, Railway Lands, Har...     0.000000   
2                                 Central Bay Street     0.014925   

   Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                0.0   0.0000              0.0000        0.0000   
1                0.0   0.0625              0.0625        0.0625   
2                0.0   0.0000              0.0000        0.0000   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0           0.000           0.0000            0.0000                  0.0   
1           0.125           0.1875            0.0625                  0.0   
2           0.000           0.0000            0.0000                  0.0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0           0.0       0.0     0.017241    0.000000                  0.0   
1           0.0       0.0     0.000000    0.000000                  0.0   
2           0.0       0.0     0.000000    0.014925                  0.0   

   Asian Restaurant  Athletics & Sports  BBQ Joint  Baby Store  Bagel Shop  \
0               0.0                 0.0   0.017241         0.0    0.017241   
1               0.0                 0.0   0.000000         0.0    0.000000   
2               0.0                 0.0   0.000000         0.0    0.000000   

     Bakery  Bank       Bar  Baseball Stadium  Basketball Stadium     Beach  \
0  0.034483   0.0  0.000000               0.0            0.017241  0.017241   
1  0.000000   0.0  0.000000               0.0            0.000000  0.000000   
2  0.000000   0.0  0.014925               0.0            0.000000  0.000000   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant    Bistro  \
0              0.0  0.034483         0.0                 0.0  0.017241   
1              0.0  0.000000         0.0                 0.0  0.000000   
2              0.0  0.000000         0.0                 0.0  0.000000   

   Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  \
0         0.0000        0.0    0.0000                   0.0        0.017241   
1         0.0625        0.0    0.0625                   0.0        0.000000   
2         0.0000        0.0    0.0000                   0.0        0.000000   

   Brewery  Bubble Tea Shop  Building  Burger Joint  Burrito Place   Butcher  \
0      0.0         0.000000       0.0      0.000000            0.0  0.017241   
1      0.0         0.000000       0.0      0.000000            0.0  0.000000   
2      0.0         0.029851       0.0      0.029851            0.0  0.000000   

   Cable Car      Café  Camera Store  Candy Store  Caribbean Restaurant  \
0        0.0  0.034483           0.0          0.0                   0.0   
1        0.0  0.000000           0.0          0.0                   0.0   
2        0.0  0.044776           0.0          0.0                   0.0   

   Cheese Shop  Chinese Restaurant  Chocolate Shop  Church  Clothing Store  \
0     0.034483            0.000000             0.0     0.0        0.017241   
1     0.000000            0.000000             0.0     0.0        0.000000   
2     0.000000            0.014925             0.0     0.0        0.000000   

   Cocktail Bar  Coffee Shop  College Arts Building  College Auditorium  \
0      0.051724     0.086207                    0.0                 0.0   
1      0.000000     0.062500                    0.0                 0.0   
2      0.000000     0.164179                    0.0                 0.0   

   College Gym  College Rec Center  Colombian Restaurant  \
0          0.0                 0.0                   0.0   
1          0.0                 0.0                   0.0   
2          0.0                 0.0                   0.0   

   Comfort Food Restaurant  Comic Shop  Concert Hall  Convenience Store  \
0                 0.017241    0.000000      0.017241                0.0   
1            

#### Let's confirm the new size

In [41]:
down_to_grouped.shape

(19, 214)

#### Let's print each neighborhood along with the top 5 most common venues

In [42]:
num_top_venues = 5

for hood in down_to_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = down_to_grouped[down_to_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
          venue  freq
0   Coffee Shop  0.09
1  Cocktail Bar  0.05
2      Beer Bar  0.03
3        Bakery  0.03
4          Café  0.03


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
             venue  freq
0  Airport Service  0.19
1   Airport Lounge  0.12
2    Boat or Ferry  0.06
3      Coffee Shop  0.06
4          Airport  0.06


----Central Bay Street----
                 venue  freq
0          Coffee Shop  0.16
1  Japanese Restaurant  0.06
2       Sandwich Place  0.06
3   Italian Restaurant  0.06
4                 Café  0.04


----Christie----
           venue  freq
0  Grocery Store  0.24
1           Café  0.18
2           Park  0.12
3     Baby Store  0.06
4    Candy Store  0.06


----Church and Wellesley----
                 venue  freq
0          Coffee Shop  0.08
1     Sushi Restaurant  0.06
2  Japanese Restaurant  0.06
3           Restaurant  0.04
4              Gay Bar  0.04


----Commerce C

#### Let's put that into a *pandas* dataframe

First, let's write a function to sort the venues in descending order.

In [46]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [48]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe

neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = down_to_grouped['Neighborhood']

for ind in np.arange(down_to_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(down_to_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
2                                 Central Bay Street           Coffee Shop   
3                                           Christie         Grocery Store   
4                               Church and Wellesley           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Cocktail Bar           Cheese Shop            Restaurant   
1        Airport Lounge              Boutique      Sculpture Garden   
2        Sandwich Place    Italian Restaurant   Japanese Restaurant   
3                  Café                  Park           Coffee Shop   
4      Sushi Restaurant   Japanese Restaurant               Gay Bar   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0    Seafood Restaurant              Beer Bar                Bakery   
1                 Plane   Rental Car Location         Boat or Ferry   
2                  Café       Thai Restaurant           Salad Place   
3    Athletics & Sports             Nightclub            Restaurant   
4            Restaurant           Yoga Studio          Burger Joint   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                  Café             Nightclub                 Museum  
1           Coffee Shop       Harbor / Marina       Airport Terminal  
2          Dessert Shop      Department Store           Burger Joint  
3                 Diner            Baby Store     Italian Restaurant  
4                 Hotel                  Café        Bubble Tea Shop

##  Cluster Neighborhoods

Run *k*-means to cluster the neighborhood into 5 clusters.

In [52]:
# set number of clusters
kclusters = 5

down_to_grouped_clustering = down_to_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(down_to_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 3, 4, 2, 0, 0, 0, 0, 0, 0], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [53]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

down_to_merged = down_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
down_to_merged = down_to_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

down_to_merged.head() # check the last columns!

Postal Code           Borough                                 Neighborhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M5E  Downtown Toronto                                  Berczy Park   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.654260 -79.360636               0           Coffee Shop   
1  43.662301 -79.389494               4           Coffee Shop   
2  43.657162 -79.378937               0        Clothing Store   
3  43.651494 -79.375418               0           Coffee Shop   
4  43.644771 -79.373306               0           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                Bakery                   Pub                  Park   
1      Sushi Restaurant           Yoga Studio                  Bank   
2           Coffee Shop    Italian Restaurant                  Café   
3                  Café          Cocktail Bar   American Restaurant   
4          Cocktail Bar           Cheese Shop            Restaurant   

  5th Most Common Venue 6th Most Common Venue      7th Most Common Venue  \
0        Breakfast Spot                  Café                    Theater   
1         Smoothie Shop              Beer Bar             Sandwich Place   
2        Cosmetics Shop       Bubble Tea Shop  Middle Eastern Restaurant   
3             Gastropub        Cosmetics Shop             Lingerie Store   
4    Seafood Restaurant              Beer Bar                     Bakery   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0           Yoga Studio          Dessert Shop             Restaurant  
1         Burrito Place                  Café                   Park  
2   Japanese Restaurant           Pizza Place       Ramen Restaurant  
3            Restaurant      Department Store         Clothing Store  
4                  Café             Nightclub                 Museum

Finally, let's visualize the resulting clusters

In [55]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(down_to_merged['Latitude'], down_to_merged['Longitude'], down_to_merged['Neighborhood'], down_to_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

##  Examine Clusters

Now, you can examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories, you can then assign a name to each cluster. I will leave this exercise to you.

#### Cluster 1

In [57]:
down_to_merged.loc[down_to_merged['Cluster Labels'] == 0, down_to_merged.columns[[1] + list(range(5, down_to_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0   Downtown Toronto               0           Coffee Shop   
2   Downtown Toronto               0        Clothing Store   
3   Downtown Toronto               0           Coffee Shop   
4   Downtown Toronto               0           Coffee Shop   
7   Downtown Toronto               0           Coffee Shop   
8   Downtown Toronto               0           Coffee Shop   
9   Downtown Toronto               0           Coffee Shop   
10  Downtown Toronto               0           Coffee Shop   
11  Downtown Toronto               0                  Café   
12  Downtown Toronto               0                  Café   
15  Downtown Toronto               0           Coffee Shop   
16  Downtown Toronto               0           Coffee Shop   
17  Downtown Toronto               0           Coffee Shop   
18  Downtown Toronto               0           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
0                 Bakery                   Pub                   Park   
2            Coffee Shop    Italian Restaurant                   Café   
3                   Café          Cocktail Bar    American Restaurant   
4           Cocktail Bar           Cheese Shop             Restaurant   
7                   Café            Restaurant          Deli / Bodega   
8               Aquarium                 Hotel                   Café   
9                  Hotel                  Café             Restaurant   
10            Restaurant                  Café                  Hotel   
11    Italian Restaurant             Bookstore                    Bar   
12           Coffee Shop    Mexican Restaurant  Vietnamese Restaurant   
15                  Café                 Hotel             Restaurant   
16           Pizza Place            Restaurant                    Pub   
17                  Café                 Hotel             Restaurant   
18      Sushi Restaurant   Japanese Restaurant                Gay Bar   

            5th Most Common Venue 6th Most Common Venue  \
0                  Breakfast Spot                  Café   
2                  Cosmetics Shop       Bubble Tea Shop   
3                       Gastropub        Cosmetics Shop   
4              Seafood Restaurant              Beer Bar   
7                             Gym       Thai Restaurant   
8                         Brewery   Sporting Goods Shop   
9             Japanese Restaurant    Italian Restaurant   
10                            Gym   American Restaurant   
11            Japanese Restaurant                Bakery   
12  Vegetarian / Vegan Restaurant                Bakery   
15                   Cocktail Bar    Italian Restaurant   
16                         Bakery    Italian Restaurant   
17                            Gym   American Restaurant   
18                     Restaurant           Yoga Studio   

        7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0                     Theater           Yoga Studio          Dessert Shop   
2   Middle Eastern Restaurant   Japanese Restaurant           Pizza Place   
3              Lingerie Store            Restaurant      Department Store   
4                      Bakery                  Café             Nightclub   
7                       Hotel             Bookstore        Clothing Store   
8                  Restaurant        Scenic Lookout   Fried Chicken Joint   
9                 Salad Place    Seafood Restaurant   American Restaurant   
10         Seafood Restaurant   Japanese Restaurant    Italian Restaurant   
11                 Restaurant              Beer Bar            Beer Store   
12              Grocery Store           Pizza Place                   Bar   
15         Seafood Restaurant              Beer Bar   Japanese Restaurant   
16                     Market                  Café         Grocery Store   
17           Asian Restaurant           Salad Place    Seafood Restaurant   
18               Burge

#### Cluster 2

In [58]:
down_to_merged.loc[down_to_merged['Cluster Labels'] == 1, down_to_merged.columns[[1] + list(range(5, down_to_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
14  Downtown Toronto               1                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
14                 Trail            Playground         Women's Store   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
14          Cupcake Shop      Doner Restaurant               Dog Run   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
14   Distribution Center        Discount Store                  Diner

#### Cluster 3

In [59]:
down_to_merged.loc[down_to_merged['Cluster Labels'] == 2, down_to_merged.columns[[1] + list(range(5, down_to_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
6  Downtown Toronto               2         Grocery Store   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
6                  Café                  Park           Coffee Shop   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
6    Athletics & Sports             Nightclub            Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
6                 Diner            Baby Store     Italian Restaurant

#### Cluster 4

In [60]:
down_to_merged.loc[down_to_merged['Cluster Labels'] == 3,down_to_merged.columns[[1] + list(range(5, down_to_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
13  Downtown Toronto               3       Airport Service   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
13        Airport Lounge              Boutique      Sculpture Garden   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
13                 Plane   Rental Car Location         Boat or Ferry   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
13           Coffee Shop       Harbor / Marina       Airport Terminal

#### Cluster 5

In [61]:
down_to_merged.loc[down_to_merged['Cluster Labels'] == 4, down_to_merged.columns[[1] + list(range(5, down_to_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
1  Downtown Toronto               4           Coffee Shop   
5  Downtown Toronto               4           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
1      Sushi Restaurant           Yoga Studio                  Bank   
5        Sandwich Place    Italian Restaurant   Japanese Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
1         Smoothie Shop              Beer Bar        Sandwich Place   
5                  Café       Thai Restaurant           Salad Place   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
1         Burrito Place                  Café                   Park  
5          Dessert Shop      Department Store           Burger Joint